[bindings] fix explog

gabrielebndn opened the pull request:
The bindings for explog were not working.
I obtained segmentation faults and/or wrong results when using fixed-size types as template parameters, so I had to switch to dynamic size (tested on my computer and on @jmirabel's laptop).

Apparently, this only happens when Pinocchio is compiled in *Release* mode. It does not happen in Debug mode, and it does not happen when neither Debug nor Release is specified.

It seems the problem is related to eigenpy. We only tested for eigenpy in Release mode.
This pull request patches the problem for explog in Pinocchio, but the problem in eigenpy should be investigated.

jcarpent commented on bindings/python/spatial/expose-explog.cpp:
```diff
@@ -15,22 +15,22 @@ namespace pinocchio
               " angular velocity during time 1.");
       
-      bp::def("Jexp3",&Jexp3_proxy<Eigen::Vector3d>,
+      bp::def("Jexp3",&Jexp3_proxy<Eigen::VectorXd>,
```
Why are you doing that: changing from Vector3d to VectorXd?

gabrielebndn replied:
That's what I explain above: it was giving wrong results with `Vector3d,` so I had to switch to `VectorXd`

jcarpent replied:
It is really weird. It is better to investigate first eigenpy rather than putting some fix here.

gabrielebndn replied:
I would rather have a working system now and think about solving this problem later. The transition on Pinocchio 2.0 is already tough enough. The correction as it stands now does not do a lot of harm, and it will be very easy to roll back.
Anyhow, you can take the newly-added lines in unittest/python/explog as a reference.
In this PR I also include a correction to python's log function.

jcarpent commented on bindings/python/spatial/explog.hpp:
```diff
@@ -23,9 +23,10 @@ namespace pinocchio
     
     template<typename Matrix3Like>
-    typename EIGEN_PLAIN_TYPE(Matrix3Like) Jlog3_proxy(const Eigen::MatrixBase<Matrix3Like> & M)
+    typename Eigen::Matrix<typename EIGEN_PLAIN_TYPE(Matrix3Like)::Scalar,3,3,EIGEN_PLAIN_TYPE(Matrix3Like)::Options>    
```
I don't know why you have to change here?

gabrielebndn replied:
I had to do this in order to be able to use `MatrixXd` as template parameter instead of `Matrix3d`, which I did for the reasons explained above

jcarpent commented on bindings/python/spatial/expose-explog.cpp:
```diff
@@ -50,7 +50,7 @@ namespace pinocchio
               " -> { v,w in se3, ||w|| < 2pi }.");
       
-      bp::def("log6",&log6<Eigen::Matrix4d>,
+      bp::def("log6",&log6<Eigen::MatrixXd>,
```
I don't think that Matrix4d is exposed. But it should be.

gabrielebndn replied:
It seems to be exposed

gabrielebndn commented:
Baiscally, with pinocchio and eigenpy both in Release mode, do this simple test:
```
v = pinocchio.utils.zero(3)
J = pinocchio.Jexp3(v)
```
The result should be the identity. Instead, I obtain something like diag([1,0.5,1.6]). This only in Release mode, because in Debug and with unspecified mode it gives the correct result

jcarpent commented:
I don't have any trouble from my side with `J = pinocchio.Jexp3(v)`.
Your issue seems to come from another place. Are you up to date with eigenpy?

gabrielebndn commented:
I am using the up-to-data installation from source. @jmirabel was also using the source installation. Are you testing in Release mode? Both Pinocchio and Eigenpy? Did you check the result is correct? @jmirabel and I were both getting the same wrong result, [[1.,0.5,0],[-0.5,0.5,0],[0,0,1.6]] if I recall.

jcarpent commented:
Yes, the result is identity.
I'm indeed using both of them in release mode.

jcarpent commented:
I got it. It comes from EigenPy and the support of `Eigen::MatrixBase<MatrixDerived>. The right destructor is not correctly invoked. I think this is related to stack-of-tasks/eigenpy#20.

gabrielebndn commented:
Then it's really weird. It's surprising @jmirabel and I are experiencing the same behavior. You may try to test the other methods too, most of them were segmenting, but at this point I expect they will work for you.

It is probably better to discard my fix. I will change the PR so to remove most fixes to the bindings, but keep the modified unit test for future reference. Do you agree?

gabrielebndn commented:
Saw your last comment. Then why is working for you, but not for us?

jcarpent commented:
It is mostly due to Eigen in fact.
Depending on your compilation flags, Eigen chooses one of the constructors (they are several for aligned data or not for instance). And the destructor should be in adequation with it. But when you compile EigenPy and Pinocchio with different options, bad things may happen. 
I already discussed this point with Eigen's main developers, but the fix I proposed was not satisfying for them.

I have suggested a while ago, to remove the shared library from eigenpy to avoid such segmentation fault. I think this is time to implement this strategy.

gabrielebndn commented:
Voilà. As #598 solved my problem, I've updated the PR.
Now basically all I'm doing is adding and fixing unit tests.
I've also fixed a little problem is python's `log` function, and slightly simplified the implementation of `log6(Matrix4d)` in C++

jcarpent commented:
looks good now! 
@gabrielebndn Thanks for this PR.
